docs/RFCS/20170920_sql_consistency_checks.md

```diff
@@ -0,0 +1,428 @@
+
+# Guide-level explanation
+
+## Terminology
+
+The **physical table** is the underlying KV pairs behind tables. This is
+where we find primary and secondary indexes (as distinct keys),
+groupings of certain data (column families), how index data is organized
+(index storing conditions), and the encoding of the data.
+
+On the other hand, the **logical table** is the SQL representation of
+the data. This is where we think about data as rows and columns. It's
+also where we have table schemas and their corresponding constraints
+(`UNIQUE`, `CHECK`, ...).
+
+Checks fall into two categories for their types. They are either a check
+on the physical table or the logical table. For example, a grouping of
+data only exist as the specific value encoded in the KV pairs
+corresponding to an SQL row, and as such the only to can access them is
+through the KV layer. Secondary indexes are an exception to this strict
+bucketing as they exist in physical tables but have been intentionally
+exposed through SQL interfaces.
+
+Alongside the check type, there are two main types of implementations of
+the checks that will be discussed in this RFC. The first is using SQL
+queries to execute the check, which can be used to check the logical
+table[0] -- hence we will call them **SQL checks**. As for the physical
+layer checks, these need to be done through operating on the KV pairs
+directly -- so we will call these **KV checks**. The method for
+running KV checks in 
+
+[0]: As an exception, secondary indexes, which exist in the physical
+table, can be checked through SQL to some extent. 
+
+## What and how
+
+Checks of both kinds can be manually invoked via the CLI to check either
+a database or table and whether to only do a specific check:
+```sql
+FSCK DATABASE <database> [FOR <checkname>];
+FSCK TABLE <table> [FOR <checkname>];
+```
+
+SQL checks can be made to run periodically and could appear as jobs.
+
+The SQL checks, as they are composed of SQL, are fairly easy to express
+and create. For example, we can check a secondary index size using[1]:
+```sql
+# Get the index size
+SELECT count(*) FROM table@index
+# Get the table size
+SELECT count(*) FROM table
+# Then assert they are equal
+```
+
+For KV checks they will be executed in distSQL processors.
+
+TODO: High level of what happens when we fail a check? (Detailed version
+below)
+
+[1]: This isn't a fully correct way to check full secondary index
+consistency. There are problems that may impact the index size for a net
+zero result, causing false positives.
+
+# Detailed design
+
+Key parts of designing the checks are:
+- How are they invoked and executed
+- How will each check be implemented
+- What do we do when a check fails
+- How we can repair various failures
+
+## Invocation and execution
+
+As mentioned earlier, all of the checks are invokable by the CLI.
+
+The checks can also be run on a schedule. The frequency can be
```

Comment:
I think there's potentially a different design doc needed here for a CRON system for cockroach which involves a cron system that is feed off a table of cron jobs